lib/widgets/login.dart

```diff
@@ -27,27 +107,57 @@ class AddAccountPage extends StatefulWidget {
 }
 
 class _AddAccountPageState extends State<AddAccountPage> {
-  final TextEditingController _controller = TextEditingController();
+  bool _inProgress = false;
+
+  final ServerUrlTextEditingController _controller = ServerUrlTextEditingController();
+  late ServerUrlParseResult _parseResult;
+
+  _serverUrlChanged() {
+    setState(() {
+      _parseResult = _controller.tryParse();
+    });
+  }
+
+  @override
+  void initState() {
+    super.initState();
+    _parseResult = _controller.tryParse();
+    _controller.addListener(_serverUrlChanged);
+  }
 
   @override
   void dispose() {
     _controller.dispose();
     super.dispose();
   }
 
-  Future<void> _onSubmitted(BuildContext context, String value) async {
-    final Uri? url = Uri.tryParse(value);
-    switch (url) {
-      case Uri(scheme: 'https' || 'http'):
-        // TODO(#35): validate realm URL further?
-        break;
-      default:
-        // TODO(#35): give feedback to user on bad realm URL
-        return;
+  Future<void> _onSubmitted(BuildContext context) async {
+    final ServerUrlParseResult(:url, :error) = _parseResult;
+    if (error != null) {
+      showErrorDialog(context: context,
+        title: 'Invalid input', message: error.message());
+      return;
     }
+    assert(url != null);
 
-    // TODO(#35): show feedback that we're working, while fetching server settings
-    final serverSettings = await getServerSettings(realmUrl: url);
+    final GetServerSettingsResult serverSettings;
+    setState(() {
+      _inProgress = true;
+    });
+    try {
+      serverSettings = await getServerSettings(realmUrl: url!);
+    } catch (e) {
+      if (!context.mounted) return;
+      // TODO(#35) give more helpful feedback; see `fetchServerSettings`
+      //   in zulip-mobile's src/message/fetchActions.js. Needs #37.
+      showErrorDialog(context: context,
+        title: 'Could not connect', message: 'Failed to connect to server: $url');
+      return;
+    } finally {
+      setState(() {
+        _inProgress = false;
```

Comment:
Another oddity I observe with the progress indicator is that when the request succeeds, the indicator gets replaced again by the icon before the navigation happens.

(I think this will still look odd if the indicator is on a separate button instead, as suggested in a previous comment.)

It looks like we can fix that by making two changes:
* Move this `finally` to the very end of the function. (This means having two nested `try` blocks — one to go with the `finally` and cover everything after setting `_inProgress = true`, and one to tightly wrap the network request and provide the `catch`.)
* Then await the `Navigator.push`.

Here that is in `git diff -b` form:
```diff
-    final GetServerSettingsResult serverSettings;
     setState(() {
       _inProgress = true;
     });
+    try {
+      final GetServerSettingsResult serverSettings;
       try {
         serverSettings = await getServerSettings(realmUrl: url!);
       } catch (e) {
         if (!context.mounted) return;
         // TODO(#35) give more helpful feedback; see `fetchServerSettings`
         //   in zulip-mobile's src/message/fetchActions.js. Needs #37.
         showErrorDialog(context: context,
           title: 'Could not connect', message: 'Failed to connect to server: $url');
         return;
-    } finally {
-      setState(() {
-        _inProgress = false;
-      });
       }
       if (context.mounted) {} // https://github.com/dart-lang/linter/issues/4007
       else {
         return;
       }
 
       // TODO(#36): support login methods beyond email/password
-    Navigator.push(context,
+      await Navigator.push(context,
           EmailPasswordLoginPage.buildRoute(serverSettings: serverSettings));
+    } finally {
+      setState(() {
+        _inProgress = false;
+      });
+    }
   }
```